File: packages/capacitor/ios/MushiMushi/Sources/MushiMushi/Queue/OfflineQueue.swift
import Foundation

/// Append-only file-backed queue that survives app restarts. Mirrors the
/// behaviour of `@mushi-mushi/core/offline-queue.ts` so the contract is
/// identical across platforms.
///
/// Threading: All file IO happens on a private serial queue. Public methods
/// are safe to call from any thread.
final class OfflineQueue {
    private let fileURL: URL
    private let maxBytes: Int
    private let queue = DispatchQueue(label: "dev.mushimushi.offline-queue")

    /// - Parameters:
    ///   - maxBytes: Soft cap for the on-disk queue. Oldest entries are
    ///     trimmed when adding would exceed this size.
    ///   - directory: Override the storage directory. Production code should
    ///     pass `nil` (defaults to Application Support); tests use a tmpdir.
    init(maxBytes: Int, directory: URL? = nil) {
        let dir = directory ?? FileManager.default
            .urls(for: .applicationSupportDirectory, in: .userDomainMask)[0]
            .appendingPathComponent("MushiMushi", isDirectory: true)
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        self.fileURL = dir.appendingPathComponent("queue.ndjson")
        self.maxBytes = maxBytes
    }

    /// Enqueue a JSON-serialisable report. Drops the oldest entries if
    /// adding this one would exceed `maxBytes`.
    func enqueue(_ payload: [String: Any]) {
        queue.sync {
            guard JSONSerialization.isValidJSONObject(payload),
                  let data = try? JSONSerialization.data(withJSONObject: payload) else {
                return
            }
            var line = data
            line.append(0x0A) // newline
            appendAndTrim(line)
        }
    }

    /// Drain up to `limit` queued reports. Caller is expected to attempt
    /// delivery and call `clearDelivered` on success.
    func peek(limit: Int) -> [[String: Any]] {
        queue.sync {
            guard let data = try? Data(contentsOf: fileURL) else { return [] }
            return data.split(separator: 0x0A)
                .prefix(limit)
                .compactMap { try? JSONSerialization.jsonObject(with: Data($0)) as? [String: Any] }
        }
    }

    /// Drop the first `count` queued reports.
    func clearDelivered(count: Int) {
        queue.sync {
            guard count > 0,
                  let data = try? Data(contentsOf: fileURL) else { return }
            var lines = data.split(separator: 0x0A)
            guard count <= lines.count else {
                try? FileManager.default.removeItem(at: fileURL)
                return
            }
            lines.removeFirst(count)
            let rebuilt = lines.flatMap { Array($0) + [UInt8(0x0A)] }
            try? Data(rebuilt).write(to: fileURL, options: .atomic)
        }
    }

    /// Number of queued reports. Cheap; counts newlines.
    var count: Int {
        queue.sync {
            guard let data = try? Data(contentsOf: fileURL) else { return 0 }
            return data.reduce(0) { $0 + ($1 == 0x0A ? 1 : 0) }
        }
    }

    private func appendAndTrim(_ line: Data) {
        var existing = (try? Data(contentsOf: fileURL)) ?? Data()
        existing.append(line)
        // Cheap newest-wins trim: drop earliest lines until under the cap.
        while existing.count > maxBytes,
              let nl = existing.firstIndex(of: 0x0A) {
            existing.removeSubrange(0...nl)
        }
        try? existing.write(to: fileURL, options: .atomic)
    }
}

File: packages/capacitor/android/mushi-android/src/main/kotlin/dev/mushimushi/sdk/storage/OfflineQueue.kt
package dev.mushimushi.sdk.storage

import android.content.Context
import com.google.gson.Gson
import java.io.File
import java.util.concurrent.locks.ReentrantLock
import kotlin.concurrent.withLock

/**
 * Append-only file-backed queue that survives app restarts. Mirrors the
 * behaviour of the iOS `OfflineQueue` and the JS core SDK so the contract
 * is identical across platforms.
 *
 * Threading: All file IO is guarded by a per-instance lock. Methods are safe
 * to call from any thread; callers should still avoid the main thread.
 */
class OfflineQueue(
    context: Context,
    private val maxBytes: Long
) {
    private val file: File = File(context.applicationContext.filesDir, "mushi/queue.ndjson").apply {
        parentFile?.mkdirs()
    }
    private val lock = ReentrantLock()
    private val gson = Gson()

    /** Enqueue a JSON-serialisable report. Drops oldest entries on overflow. */
    fun enqueue(payload: Map<String, Any?>) = lock.withLock {
        val json = gson.toJson(payload).toByteArray(Charsets.UTF_8)
        val combined = ByteArray(json.size + 1).apply {
            System.arraycopy(json, 0, this, 0, json.size)
            this[json.size] = 0x0A
        }
        var existing = if (file.exists()) file.readBytes() else ByteArray(0)
        existing = existing + combined
        while (existing.size > maxBytes) {
            val nl = existing.indexOfFirst { it == 0x0A.toByte() }
            if (nl < 0) break
            existing = existing.copyOfRange(nl + 1, existing.size)
        }
        file.writeBytes(existing)
    }

    /** Drain up to [limit] queued reports without removing them. */
    fun peek(limit: Int): List<Map<String, Any?>> = lock.withLock {
        if (!file.exists()) return@withLock emptyList()
        file.readLines(Charsets.UTF_8)
            .asSequence()
            .filter { it.isNotBlank() }
            .take(limit)
            .mapNotNull { line ->
                runCatching {
                    @Suppress("UNCHECKED_CAST")
                    gson.fromJson(line, Map::class.java) as? Map<String, Any?>
                }.getOrNull()
            }
            .toList()
    }

    /** Drop the first [count] queued reports. */
    fun clearDelivered(count: Int) = lock.withLock {
        if (count <= 0 || !file.exists()) return@withLock
        val remaining = file.readLines(Charsets.UTF_8)
            .filter { it.isNotBlank() }
            .drop(count)
        if (remaining.isEmpty()) {
            file.delete()
        } else {
            file.writeText(remaining.joinToString("\n", postfix = "\n"), Charsets.UTF_8)
        }
    }

    /** Number of queued reports. */
    fun count(): Int = lock.withLock {
        if (!file.exists()) 0
        else file.readLines(Charsets.UTF_8).count { it.isNotBlank() }
    }
}
